Improve pandas/xarray/... conversion

oscargus opened the pull request:
## PR Summary

See https://stackoverflow.com/questions/13187778/convert-pandas-dataframe-to-numpy-array/54508052#54508052 for motivation.

Related to #16402

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->
**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

**Documentation**
- [N/A] New features are documented, with examples if plot related.
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of main, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

greglucas commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -7907,8 +7907,8 @@ def violinplot(self, dataset, positions=None, vert=True, widths=0.5,
-            if hasattr(X, 'values'):  # support pandas.Series
-                X = X.values
+            if hasattr(X, 'to_numpy'):  # support pandas.Series
+                X = X.to_numpy()
```
Are there any libraries we need to worry about that implement `X.values`, but not `X.to_numpy()`? Maybe we also want to leave an `elif` with the `X.values` block as a fallback just in case...?

QuLogic commented:
Perhaps should be coordinated with #22141?

oscargus commented:
> Perhaps should be coordinated with #22141?

That's where I got the idea. As seen, I have not touched `cbook`.

Related to the question from @greglucas, I do not know. But I saw this code in `cbook` (which I didn't touch):
https://github.com/matplotlib/matplotlib/blob/3a994d25f00e40df18fefc4c8b2ba57519bebacc/lib/matplotlib/cbook/__init__.py#L1369-L1377

This is a bit more complex as it uses a try-except approach (not sure how much that affects things though), has a fallback on values and check that values actually returns an nparray. One can of course use a similar approach here (and in #22141). Possibly slightly improved as I do not know if there are cases if values actually is non-trivial, so no need to run it twice.

greglucas commented:
I guess you could write a `cbook._unpack_pandas(x)` helper to use that in all of these areas for consistency? I would say it is safer to keep the `.values` fallback, and if you do want to get rid of it eventually, you could put in a deprecation warning if that branch is hit.

oscargus commented:
Yes, I was also thinking about a helper function (but didn't know where to place it, so great info that cbook is the place). I do not have any strong opinions as such, more that I read the link and it seemed like the right thing to do.

oscargus commented:
Seems like the easiest way is to wait for #22141, add those conversions here as well, and then discuss the correct name for the function.

oscargus commented:
I'm thinking that maybe `_unpack_to_numpy` is a general enough name?

oscargus commented:
This is now updated:
* renamed function to `_unpack_to_numpy`
* added test for xarray (now an extra test dependency)
* (removed duplicate test code for pandas)

tacaswell commented:
Please update the lines touched by #22141

oscargus commented:
There is also this function where it doesn't work to simply replace `to_numpy` with `_unpack_to_numpy`. No idea why. (One reason may be that there is no attribute error for `to_numpy` anymore, so some types of input data is handled well by `_check_1d`.)

https://github.com/matplotlib/matplotlib/blob/03598322b99b11d8c59716d3c67450852945333a/lib/matplotlib/cbook/__init__.py#L1607-L1639

greglucas commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2231,3 +2223,20 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
+    if hasattr(x, 'to_numpy'):
+        # Assume that any function to_numpy() do actually return a numpy array
+        return x.to_numpy()
+    if hasattr(x, 'values'):
```
I think you can simplify the nested ifs and temporary array here?

```suggestion
    if isinstance(getattr(x, 'values', None), np.ndarray):
        return x.values
```

oscargus replied:
Oh, I think the idea was to avoid calling values several times in case the generation was non-trivial.

greglucas replied:
Good point, thanks for clarifying!

greglucas commented on lib/matplotlib/tests/test_cbook.py:
```diff
@@ -680,14 +680,37 @@ def test_reshape2d_pandas(pd):
+    np.testing.assert_array_equal(Idx, IdxRef)
+
+
+def test_index_of_xarray(xr):
```
Does xarray get us more coverage here? They have a `to_numpy()` method the same as pandas I believe.
https://xarray.pydata.org/en/stable/generated/xarray.DataArray.to_numpy.html

So, it seems like a pretty heavy dependency to add for just this one test...

oscargus replied:
Not so much for coverage as for actually testing using data of specified formats. With the discussion about which formats we support, it makes sense to test them as well. Right now some of these are tested in the plots, but it can possibly make sense to simply test them here as these are the core function used to get data that can be plotted. 

If we claim (which we actually don't, maybe we should?) that we can plot xarray, we should probably test it as well. And other types that we may want to claim to support. Or maybe fork off a specific dependency test that is not executed on all platforms/version, including pandas (which is 11.7 MB, xarray is 870 kB).

(There is another xarray-test above, so two.)

I can of course remove them, but I think we should discuss if we want to support more formats than pandas and numpy (and Python list/tuple), and, if so, have explicit tests for them.

greglucas replied:
I agree, we should probably discuss what we want to support/test. To me, this doesn't seem to add a whole lot of value for adding a new dependency.

There was also a discussion around removing Scipy as a  dependency in the docs: https://github.com/matplotlib/matplotlib/pull/22120

tacaswell commented:
I made an executive decision to install xarray on CI.  

We already have all of its dependencies installed and it is a pure-python package.

timhoffm commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -2231,3 +2223,20 @@ def _picklable_class_constructor(mixin_class, fmt, attr_name, base_class):
+
+
+def _unpack_to_numpy(x):
+    """Internal helper to extract data from e.g. pandas and xarray objects."""
```
Please document what we intend to support, i.e. everything with .to_numpy() or .values, and what types we expect to catch with it, e.g. values-> older pandas dataframes(?).

tacaswell commented:
We discussed this on the call, I will write up an issues shortly about the documentation of the intention etc.

QuLogic commented:
meeseeksdev backport to v3.5.x
